File: portal/omdb.py
"""OMDb adapter. Never expose upstream errors/URLs containing the API key."""
import hashlib
import json
import math
import re
from urllib.parse import urlsplit

import requests
from django.conf import settings
from django.core.cache import caches

IMDB_ID = re.compile(r'tt[0-9]{7,10}')


class MovieApiError(Exception):
    def __init__(self, status, message):
        self.status = status
        super().__init__(message)


def text(value):
    return value if isinstance(value, str) and value != 'N/A' else 'Нет данных'


def summary(value):
    if not isinstance(value, dict) or not isinstance(value.get('imdbID'), str) or not IMDB_ID.fullmatch(value['imdbID']) or not isinstance(value.get('Title'), str):
        raise MovieApiError(502, 'OMDb вернул некорректную карточку фильма.')
    poster = None
    try:
        url = urlsplit(value.get('Poster', ''))
        if url.scheme == 'https' and url.hostname in ('m.media-amazon.com', 'ia.media-imdb.com') and not url.username and not url.password:
            poster = value['Poster']
    except (ValueError, TypeError, AttributeError):
        pass
    return {'id': value['imdbID'], 'title': value['Title'], 'year': text(value.get('Year')), 'poster': poster}


def request(params):
    if not settings.OMDB_API_KEY.strip():
        raise MovieApiError(503, 'Поиск не настроен: добавьте OMDB_API_KEY и перезапустите сервер.')
    cache_key = hashlib.sha256(json.dumps(params, sort_keys=True).encode()).hexdigest()
    cached = caches['omdb'].get(cache_key)
    if cached is not None:
        return cached
    try:
        response = requests.get('https://www.omdbapi.com/', params={**params, 'apikey': settings.OMDB_API_KEY, 'r': 'json'}, timeout=8, allow_redirects=False)
        if response.status_code in (401, 403):
            raise MovieApiError(503, 'Ключ OMDb недействителен или ещё не активирован.')
        if response.status_code == 429:
            raise MovieApiError(503, 'Лимит OMDb исчерпан. Повторите запрос позже.')
        if response.status_code != 200:
            raise MovieApiError(502, 'OMDb временно недоступен. Попробуйте позже.')
        data = response.json()
    except requests.Timeout:
        raise MovieApiError(504, 'OMDb не ответил вовремя. Повторите запрос.') from None
    except (requests.RequestException, ValueError):
        raise MovieApiError(502, 'Не удалось связаться с OMDb. Повторите запрос позже.') from None
    if not isinstance(data, dict):
        raise MovieApiError(502, 'OMDb вернул некорректный ответ.')
    if data.get('Response') == 'False':
        message = str(data.get('Error', ''))
        for pattern, status, safe in [
            ('not found|incorrect imdb', 404, 'Фильм не найден.'),
            ('too many results', 400, 'Слишком много результатов. Уточните название или год.'),
            ('limit', 503, 'Дневной лимит OMDb исчерпан. Попробуйте завтра.'),
            ('key|activat', 503, 'Ключ OMDb недействителен или ещё не активирован.'),
        ]:
            if re.search(pattern, message, re.I):
                raise MovieApiError(status, safe)
        raise MovieApiError(502, 'OMDb не смог выполнить запрос.')
    if data.get('Response') != 'True':
        raise MovieApiError(502, 'OMDb вернул некорректный ответ.')
    caches['omdb'].set(cache_key, data, 600)
    return data


def search(query, year, page):
    try:
        data = request({'s': query, 'type': 'movie', 'page': str(page), **({'y': year} if year else {})})
    except MovieApiError as error:
        if error.status == 404:
            return {'movies': [], 'totalResults': 0, 'page': page, 'totalPages': 0}
        raise
    if not isinstance(data.get('Search'), list) or not re.fullmatch(r'[0-9]+', str(data.get('totalResults', ''))):
        raise MovieApiError(502, 'OMDb вернул некорректный список фильмов.')
    total = int(data['totalResults'])
    return {'movies': [summary(v) for v in data['Search']], 'totalResults': total, 'page': page, 'totalPages': min(100, math.ceil(total / 10))}


def details(imdb_id):
    data = request({'i': imdb_id, 'plot': 'full', 'type': 'movie'})
    return {**summary(data), **{dest: text(data.get(source)) for dest, source in {
        'plot': 'Plot', 'genre': 'Genre', 'director': 'Director', 'actors': 'Actors',
        'runtime': 'Runtime', 'rating': 'imdbRating', 'country': 'Country',
    }.items()}}

File: server/omdb.ts
import { Router } from 'express'
import type { MovieDetails, MovieSearch, MovieSummary } from '../src/types/movies.js'

type JsonRecord = Record<string, unknown>
type Fetcher = typeof fetch
const imdbIdPattern = /^tt\d{7,10}$/
const posterHosts = new Set(['m.media-amazon.com', 'ia.media-imdb.com'])

export class MovieApiError extends Error {
  constructor(public status: number, message: string) {
    super(message)
  }
}

const record = (value: unknown): JsonRecord => {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new MovieApiError(502, 'OMDb вернул некорректный ответ.')
  }
  return value as JsonRecord
}

const text = (value: unknown): string => typeof value === 'string' && value !== 'N/A' ? value : 'Нет данных'

const posterUrl = (value: unknown): string | null => {
  if (typeof value !== 'string') return null
  try {
    const url = new URL(value)
    return url.protocol === 'https:' && posterHosts.has(url.hostname) && !url.username && !url.password ? url.href : null
  } catch {
    return null
  }
}

const summary = (value: unknown): MovieSummary => {
  const movie = record(value)
  if (typeof movie.imdbID !== 'string' || !imdbIdPattern.test(movie.imdbID) || typeof movie.Title !== 'string') {
    throw new MovieApiError(502, 'OMDb вернул некорректную карточку фильма.')
  }
  return { id: movie.imdbID, title: movie.Title, year: text(movie.Year), poster: posterUrl(movie.Poster) }
}

export const createOmdbClient = ({
  apiKey,
  fetcher = fetch,
  timeoutMs = 8000,
  now = Date.now,
}: { apiKey?: string; fetcher?: Fetcher; timeoutMs?: number; now?: () => number }) => {
  // Кэш не содержит ключ, ограничен 100 ответами и живёт только в памяти процесса.
  const cache = new Map<string, { expiresAt: number; data: JsonRecord }>()

  const request = async (params: Record<string, string>): Promise<JsonRecord> => {
    if (!apiKey?.trim()) throw new MovieApiError(503, 'Поиск не настроен: добавьте OMDB_API_KEY в .env и перезапустите сервер.')
    const cacheKey = new URLSearchParams(params).toString()
    const cached = cache.get(cacheKey)
    if (cached && cached.expiresAt > now()) return cached.data
    cache.delete(cacheKey)

    const url = new URL('https://www.omdbapi.com/')
    url.search = new URLSearchParams({ ...params, apikey: apiKey, r: 'json' }).toString()
    let data: JsonRecord
    try {
      const response = await fetcher(url, { signal: AbortSignal.timeout(timeoutMs), redirect: 'error' })
      if (response.status === 401 || response.status === 403) throw new MovieApiError(503, 'Ключ OMDb недействителен или ещё не активирован.')
      if (response.status === 429) throw new MovieApiError(503, 'Лимит OMDb исчерпан. Повторите запрос позже.')
      if (!response.ok) throw new MovieApiError(502, 'OMDb временно недоступен. Попробуйте позже.')
      data = record(await response.json())
    } catch (error) {
      if (error instanceof MovieApiError) throw error
      if (error instanceof Error && (error.name === 'TimeoutError' || error.name === 'AbortError')) {
        throw new MovieApiError(504, 'OMDb не ответил вовремя. Повторите запрос.')
      }
      // Не выводим исходную ошибку fetch: она может содержать URL с ключом.
      throw new MovieApiError(502, 'Не удалось связаться с OMDb. Повторите запрос позже.')
    }

    if (data.Response === 'False') {
      const message = typeof data.Error === 'string' ? data.Error : ''
      if (/not found|incorrect imdb/i.test(message)) throw new MovieApiError(404, 'Фильм не найден.')
      if (/too many results/i.test(message)) throw new MovieApiError(400, 'Слишком много результатов. Уточните название или год.')
      if (/limit/i.test(message)) throw new MovieApiError(503, 'Дневной лимит OMDb исчерпан. Попробуйте завтра.')
      if (/key|activat/i.test(message)) throw new MovieApiError(503, 'Ключ OMDb недействителен или ещё не активирован.')
      throw new MovieApiError(502, 'OMDb не смог выполнить запрос.')
    }
    if (data.Response !== 'True') throw new MovieApiError(502, 'OMDb вернул некорректный ответ.')
    if (cache.size >= 100) cache.delete(cache.keys().next().value!)
    cache.set(cacheKey, { data, expiresAt: now() + 10 * 60 * 1000 })
    return data
  }

  return {
    async search(query: string, year: string, page: number): Promise<MovieSearch> {
      let data: JsonRecord
      try {
        data = await request({ s: query, type: 'movie', page: String(page), ...(year ? { y: year } : {}) })
      } catch (error) {
        if (error instanceof MovieApiError && error.status === 404) return { movies: [], totalResults: 0, page, totalPages: 0 }
        throw error
      }
      if (!Array.isArray(data.Search) || !/^\d+$/.test(String(data.totalResults))) {
        throw new MovieApiError(502, 'OMDb вернул некорректный список фильмов.')
      }
      const totalResults = Number(data.totalResults)
      return { movies: data.Search.map(summary), totalResults, page, totalPages: Math.min(100, Math.ceil(totalResults / 10)) }
    },
    async details(id: string): Promise<MovieDetails> {
      const data = await request({ i: id, plot: 'full', type: 'movie' })
      return {
        ...summary(data), plot: text(data.Plot), genre: text(data.Genre), director: text(data.Director),
        actors: text(data.Actors), runtime: text(data.Runtime), rating: text(data.imdbRating), country: text(data.Country),
      }
    },
  }
}

export const createMovieRouter = (client: ReturnType<typeof createOmdbClient>): Router => {
  const router = Router()

  router.get('/', async (request, response, next) => {
    const query = typeof request.query.query === 'string' ? request.query.query.trim() : ''
    const year = request.query.year === undefined ? '' : request.query.year
    const page = request.query.page === undefined ? '1' : request.query.page
    if (query.length < 2 || query.length > 100 || typeof year !== 'string' || (year !== '' && !/^(18|19|20|21)\d{2}$/.test(year)) || typeof page !== 'string' || !/^[1-9]\d{0,2}$/.test(page) || Number(page) > 100) {
      response.status(400).json({ ok: false, message: 'Название: 2–100 символов; год: 1800–2199; страница: 1–100.' })
      return
    }
    try {
      response.json({ ok: true, ...await client.search(query, year, Number(page)) })
    } catch (error) { next(error) }
  })

  router.get('/:id', async (request, response, next) => {
    if (!imdbIdPattern.test(request.params.id)) {
      response.status(400).json({ ok: false, message: 'Неверный IMDb ID: ожидается tt и 7–10 цифр.' })
      return
    }
    try {
      response.json({ ok: true, movie: await client.details(request.params.id) })
    } catch (error) { next(error) }
  })
  return router
}
